### ISSUE TITLE
http/2 headers updates discussion

### ISSUE THREAD

**Issue by Scottmitch** [netty/netty#3503]
1. Investigation is needed into the usage of the `String` class with respect to headers.  For example [sensitiveHeaders](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersEncoder.java#L41) is using `String` but may be better off using `AsciiString`.

2. It may be best to provide an interface to determine if the header should be `sensitive` or not instead of forcing a map lookup for every header value.  The default implementation could just do a map lookup (which would be whatever map we create by default) but the extension point would allow the user to do what they please.

3. Since HPACK is *almost* an RFC it may be worth discussing if making a fork of twitter's HPACK implementation that is more friendly to Netty is worth it.  For example using `ByteBuf` instead of `byte[]` would be an improvement.  We could also investigate if alternatives to `InputStream` and `OutputStream` would be an improvement.

**Comment by Scottmitch** [comment]
@nmittler - FYI

**Comment by Scottmitch** [comment]
@jpinner - Perhaps we can chat offline but I am wondering what your use case for  https://github.com/twitter/hpack is and if for example you would be open to the idea of allowing the interfaces to be more friendly to Netty :) (using ByteBuf, etc...). @nmittler - I'm assuming if there is a discussion you would also be interested?

**Comment by nmittler** [comment]
@louiscryan may also be interested

+1 it would be great to support `ByteBuf`

**Comment by louiscryan** [comment]
interfaces would be useful for a number of reasons though we might need a more refined interface than just bytebuf. A good start would be to get some JMH benchmarks written up for hpack to exercise it

**Comment by nmittler** [comment]
@Scottmitch I think you have the benchmarks for twitter's hpack now ... can you summarize your findings in this issue?

**Comment by Scottmitch** [comment]
@nmittler @louiscryan - The results I collected from the benchmark can be viewed here: https://github.com/netty/netty/pull/3392#issuecomment-83735958. I think we could have a more comprehensive benchmark of the headers with the following:
1. The headers sizes were more or less "randomly" chosen. We should do some analysis of what are "representational" sizes of headers.  This includes how many name/value pairs, and also the size of each name and value.
2. Currently header names/values are randomly generated.  It may be good to use "real world" header names/values to take advantage of the static tables more.
3. One which exercises the `sensitive` aspect of HPACK.  The benchmark just takes the default (all are not-sensative).
4. More isolation on the headers interface.

That being said there is a digression of performance as header size (name/value) increases.  The drop off in performance may make sense because more headers means more work for HPACK. The expectation is that HPACK should be indexing all non-sensative headers and keep the write overhead relatively similar over all sizes of benchmarks (assuming header name/values remain the same), but the comparison overhead will go up.

**Comment by Scottmitch** [comment]
@nmittler - I'm working on a PR for the first 2 points of this issue.  I may move modifying/forking twitter's HPACK to a separate issue...

**Comment by Scottmitch** [comment]
PR for https://github.com/netty/netty/pull/3578 addresses parts 1 and 2 of this issue. Closing this issue and we can track HPACK updates under another issue.

**Comment by Scottmitch** [comment]
See https://github.com/netty/netty/issues/3503 for HPACK investigation.
